spinalcordtoolbox/math.py

```diff
@@ -113,14 +106,61 @@ def dilate(data, size, shape, dim=None):
     """
     if isinstance(data, Image):
         im_out = data.copy()
-        im_out.data = dilate(data.data, size, shape, dim)
+        im_out.data = dilate(data.data, size, shape, dim, islabel)
         return im_out
     else:
         footprint = _get_footprint(shape, size, dim)
-        if data.dtype in ['uint8', 'uint16']:
-            return rank.maximum(data, footprint=footprint)
+        if islabel:
+            return _dilate_point_labels(data, footprint=footprint)
         else:
-            return dilation(data, footprint=footprint, out=None)
+            if data.dtype in ['uint8', 'uint16']:
+                return rank.maximum(data, footprint=footprint)
+            else:
+                return dilation(data, footprint=footprint)
+
+
+def _dilate_point_labels(data, footprint):
+    """
+    A more efficient dilation algorithm when we know the image is mostly zero (i.e. point label image)
+    """
+    if len(data.shape) != len(footprint.shape):
+        raise ValueError(f"incompatible shapes: {data.shape=} {footprint.shape=}")
+    if footprint.size == 0:
+        raise ValueError(f"cannot handle empty footprint: {footprint.shape=}")
+
+    data_shape = np.array(data.shape)
+    fp_shape = np.array(footprint.shape)
+
+    # For odd-length dimensions, the center of the footprint is unique.
+    # For even-length dimensions, we round down.
+    fp_center = (fp_shape - 1) // 2
+
+    data_out = np.zeros_like(data)
+    for data_coords in np.argwhere(data):
+        # data_coords is a numpy array, which behaves differently from tuples
+        # when indexing another numpy array. We want the tuple behaviour.
+        data_value = data[tuple(data_coords)]
+
+        # If a data coordinate is too close to the edge, we have to use a
+        # truncated version of the footprint. These formulas compute the
+        # adjusted slice start and stop coordinates in the data array, and in
+        # the footprint array.
+        pre_distances = np.min(data_coords, fp_center)
+        data_start = data_coords - pre_distances
+        fp_start = fp_center - pre_distances
+
+        post_distances = np.min(data_shape - data_coords, fp_shape - fp_center)
+        data_stop = data_coords + post_distances
+        fp_stop = fp_center + post_distances
+
+        data_loc = tuple(slice(start, stop) for start, stop in zip(data_start, data_stop))
+        fp_loc = tuple(slice(start, stop) for start, stop in zip(fp_start, fp_stop))
```

Comment:
I'll admit that I do find this specific organization of things a little unclear (it's hard for me to conceptualize what the variables are supposed to represent when just scanning through).

That said, once I get to the `np.maximum.at` part, it becomes a lot more clear. So, I think I'd maybe like to refactor it just a little (for clarity purposes only)?

I'd start with a comment that clarifies what our end goal is:

```python
        # To dilate each voxel, we multiply the footprint by the voxel's value
        # 
        # But, to preserve overlapping dilated voxels, we'd like to use `np.maximum` 
        # to take the maximum (so that the '0' voxels in the footprint don't overwrite
        # any existing nonzero voxels in the image).
        # 
        # To perform `np.maximum` in place, we use the `np.ufunc.at` syntax, which
        # requires an `indices` argument to specify *where* to perform the maximum at.
        # So, we just need to compute the bounding box surrounding the voxel, 
        # specified as [(x1,x2), (y1,y2), (z1,z2)]
```

Then, for the `np.maximum.at` call, we might explicitly specify the argument names to tie back into the earlier comment:

```python
        np.maximum.at(a=data_out, indices=data_loc, b=data_value*footprint[fp_loc])
```

And finally, I might change the `pre_`/`post_distances` blocks to better convey what values are being computed? These two lines in particular are really tripping me up for some reason, I think because of the implied "`- 0`" that's missing from the first statement:

```python
        pre_distances = np.min(data_coords, fp_center)
        post_distances = np.min(data_shape - data_coords, fp_shape - fp_center)
```

I find it really hard to parse what's happening here? Looking it over, I *think* this is what's happening:

```python
        # Make sure that the footprint doesn't exceed the image boundaries
        # If the footprint fits inside the image, then:
        #    - (width of footprint) <= (distance from voxel to image boundary) 
        # However, if the footprint would go outside the boundaries, then:
        #    - (distance from voxel to image boundary) < (width of footprint)
        # So we take the minimum to always use the shorter of the two distances.
```

One last thing that could be done to make it a little easier to grok is to nix the "pre" and "post" variable names? I totally get why you did it... there's not really an elegant word for "left/right" (since we're covering x/y/z), and "plus/minus" doesn't quite work either, so "pre/post" feel the best? 

But maybe something like this could make more sense (along with adding the redundant `- 0` to align the subtractions, and make it clear we're getting the distance from 0 to the center).

All together, maybe this would look like:

```python
        # Make sure that the footprint doesn't exceed the image boundaries
        # If the footprint fits inside the image, then:
        #    - (width of footprint) <= (distance from voxel to image boundary) 
        # However, if the footprint would go outside the boundaries, then:
        #    - (distance from voxel to image boundary) < (width of footprint)
        # So we take the minimum to always use the shorter of the two distances.
        distances = (
            # Distance from 0 --> voxel
            np.min(data_coords - (0, 0, 0),
                   fp_center - (0, 0, 0)),
            # Distance from voxel --> data boundaries
            np.min(data_shape - data_coords,
                   fp_shape - fp_center)
        )
        data_start = data_coords - distances[0]
        data_stop  = data_coords + distances[1]
        fp_start = fp_center - distances[0]
        fp_stop  = fp_center + distances[1]
```